### PR TITLE
Add iterators over hypercube vertices and edges

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

These iterators help looping over all corners and edges of a cube. I'm using them to construct weighting functions for my Schwarz subdomain solver.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Domain/Structure/VertexIterator.hpp
Comment (moxcodes) [inline src/Domain/Structure/VertexIterator.hpp]: I feel these structures  should have significantly more description about their intended use and member functions.

How do you plan to use these? Most iterators in the STL and in spectre have either dereference operators, implicit conversion operators, or member functions to output the full state information so that can be passed easily into functions or objects that depend on that state. These currently have the appearance of requiring `Dim` calls to `side` to get complete state information, which seems cumbersome to me intuitively, but perhaps I'm not properly understanding your use case.

Comment (nilsvu) [reply src/Domain/Structure/VertexIterator.hpp]: Thanks for reviewing @moxcodes! Here's some "pseudocode" that illustrates how I currently use these iterators in the Schwarz weighting functions:

```cpp
// Computing weights on subdomains that overlap across corners and edges
for (VertexIterator<2> corner{}; corner; ++corner) {
  // Skip corners to external boundaries
  if (skip_corner(corner)) {
    continue;
  }
  for (size_t d = 0; d < 2; ++d) {
    get(corner_weight) *= intruding_weight(
        logical_coords.get(d), overlap_widths[d], corner.side(d));
  }
  // ...
}
for (EdgeIterator<2> edge{}; edge; ++edge) {
  const size_t d = edge.dimension();
  const Side side = edge.side(0);
  // Skip edges to external boundaries
  if (skip_edge(edge)) {
    continue;
  }
  get(edge_weight) *= intruding_weight(
      logical_coords.get(d), overlap_widths[d], side);
  // ...
}
```

These iterators are similar to the `SliceIterator` and `StripeIterator`. They encode the current state in the iterator object, which can be passed around (e.g. to the `skip_corner` and `skip_edge` functions).

Do you think these hypercube iterators can be useful, or should I inline their logic where I iterate over corners or edges in the Schwarz code? I'm not very attached to this code, but thought it could be a nice re-usable way of abstracting away the hypercube geometry. I'm happy to hear your suggestions!

Comment (moxcodes) [reply src/Domain/Structure/VertexIterator.hpp]: I do think that this sort of iterator can be useful -- from the code snippet, the only improvement that I'd suggest is try to construct the iterator in such a fashion that it could be used with range-based iteration (perhaps with some type of begin/end factory functions for the iterator?).

The edge part of this code seems counter-intuitive to me -- it has the appearance that if I have a consistent way of ordering the sides outside of the edge iterator, the edge iterator means different things as it advances (because it computes the upper/lower via a bitset of length dim-1) -- so as the iterator advances, if I'm interested in a particular side index, I will need to decrement it or omit the query to `edge.side` depending on what `edge.dimension` is.
Would it make the interface more difficult for your use-case if the `edge.side` function took the full range of dimensions and returned a `boost::optional`, giving `boost::none` if the side requested is invalid for the current edge?

**Comment by nilsvu** [comment]
@moxcodes I addressed your feedback and ended up re-writing these iterators as a single class template that works for any k-dimensional element on the boundary of an n-dimensional hypercube. The iterator now has an `operator*` that returns a state-encoding vertex/edge/... as you suggested, and it works with range-based loops. To address your other concern I added a `side_in_parent_dimension(size_t)` function that works with the dimension of the hypercube, instead of the dimension of the edge(/face/...). Could you take another look and let me know what you think?

**Comment by nilsvu** [comment]
@wthrowe could you perhaps do the second review?

File: src/Domain/Structure/Hypercube.hpp
Comment (moxcodes) [inline src/Domain/Structure/Hypercube.hpp]: probably want `operator==` for the iterator too.

File: src/Domain/Structure/Hypercube.hpp
Comment (moxcodes) [inline src/Domain/Structure/Hypercube.hpp]: It's generally nice to define the prefix operator as well (denoted as `operator++(int)`), and that one can return by non-const reference

File: src/Domain/Structure/Hypercube.hpp
Comment (moxcodes) [inline src/Domain/Structure/Hypercube.hpp]: It's a little unconventional to have `begin` in the iterator class itself, I think, but I think it does make some sense for this case. But can the `begin()` and `end()` be static as they act as 'factory' methods for a fresh iterator?

**Review by moxcodes — COMMENTED** [review]
Looks good, I like the generic utility. I just have a couple of tweaks to suggest for the iterator interface.

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.hpp:58]: You have a lot of SFINAE constructor overloads here, and I think most of them are just saving you from having to enclose your arguments in double braces.  Is there a reason you need them?

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.hpp:137]: If you intend to use this for high dimensions, this should be written in a way less susceptible to overflow.  My guess is you don't expect to use it that way, in which case this is fine.

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [reply src/Domain/Structure/Hypercube.hpp]: To be (hopefully) a bit clearer here, the prefix operator (`operator++()`) should return a non-const reference, and a postfix operator (`operator++(int)`) should return by value.

File: tests/Unit/Domain/Structure/Test_Hypercube.cpp
Comment (wthrowe) [inline tests/Unit/Domain/Structure/Test_Hypercube.cpp]: `Edge<2>{...` for consistency.

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.hpp:46]: Test this.

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.hpp:62]: Test this.

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.hpp:66]: Test this.

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.hpp:72]: Test this.

And so on.  Make sure everything's tested.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/Structure/Hypercube.hpp
Comment (nilsvu) [reply src/Domain/Structure/Hypercube.hpp]: It's not required for the an iterator. I would implement it as `return not(*this != rhs)`, which is ugly enough that I didn't do it. Do you think I should add it nonetheless?

File: src/Domain/Structure/Hypercube.hpp
Comment (nilsvu) [reply src/Domain/Structure/Hypercube.hpp:58]: No reason other than convenience and a nicer interface. Take an `Edge` for example. It has only one `dim_in_parent`, no matter the dimension of the hypercube it's embedded in, so wrapping the `dim_in_parent` in double braces felt silly. The other constructor I could remove is `Vertex(Indices... indices)`. It just eliminates the double braces around the indices. Would you prefer I remove either, or both?

File: src/Domain/Structure/Hypercube.hpp
Comment (nilsvu) [reply src/Domain/Structure/Hypercube.hpp:137]: I'm only instantiating the template for `HypercubeDim<=3`, that's probably all we need.

**Comment by nilsvu** [comment]
@moxcodes @wthrowe thanks for your reviews, fixups posted!

File: src/Domain/Structure/Hypercube.hpp
Comment (moxcodes) [reply src/Domain/Structure/Hypercube.hpp]: I agree it's not strictly required for the standard iterator use cases, and the standard seems slightly unclear, but I think in terms of interface design it's best to avoid the 'surprising' choices of defining either `operator==` or `operator!=` without the complimentary operator.

**Comment by moxcodes** [comment]
I'm happy for you to squash (along with the suggested equality operator), pending @wthrowe 's go-ahead.

**Comment by nilsvu** [comment]
@wthrowe Could you check the fixups please?

File: src/Domain/Structure/Hypercube.hpp
Comment (wthrowe) [reply src/Domain/Structure/Hypercube.hpp:58]: I would drop both, but I won't object if you leave them unless someone else complains.

File: src/Domain/Structure/Hypercube.cpp
Comment (wthrowe) [inline src/Domain/Structure/Hypercube.cpp]: I don't understand this comment.  ("An iterator only requires the inequality operator"?  Input iterators certainly need both.)  Just do comparisons the same way they are done everywhere else.  Then you don't have to explain it.

Also, binary operators should be free functions unless they access private state.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by nilsvu** [comment]
@wthrowe pushed a fixup for the `operator==`, could you take another look?

**Comment by wthrowe** [comment]
`!=` should be a free function.  And actually I think we've been doing `==` as a friend function, rather than a member.

**Comment by nilsvu** [comment]
@wthrowe right of course, fixed

**Comment by wthrowe** [comment]
Squash.

**Review by wthrowe — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]
Looks good to me

**Comment by nilsvu** [comment]
Thanks for reviewing @moxcodes and @wthrowe! Seems ready to merge @wthrowe

**Pull request merged**
